packages/client/src/pages/Game/engine/TDEngine.tsx

```diff
@@ -0,0 +1,770 @@
+    public waveGenerator: ITDEngine["waveGenerator"] = null,
+  ) {
+    this.setMap(new Map(this));
+    this.waveGenerator = new WaveGenerator(this);
+    this.money = this.initialGameParams.money;
+    this.lives = this.initialGameParams.lives;
+  }
+
+  public restartGame() {
+    clearInterval(this.waveGenerator!.waveCountdownTimer!);
+    clearTimeout(this.waveGenerator!.waveTimerBetweenWaves!);
+    this.waveGenerator!.waveCountdownTimer = null;
+    this.waveGenerator!.waveTimerBetweenWaves = null;
+    this.waveGenerator!.isInitialized = false;
+    this.enemies = [];
+    this.towers = [];
+    this.projectiles = [];
+    // create mapTilesArr
+    this.map?.createMapTilesArr();
+    // pop tiles which is occupied by map path
+    this.map?.popMapPathTiles();
+    // game params
+    this.money = this.initialGameParams.money;
+    this.lives = this.initialGameParams.lives;
+    this.score = 0;
+    // spawner
+    this.waveGenerator!.waveParams.currentWave = 1;
+    this.waveGenerator!.waveCountdown = Math.floor(
+      this.waveGenerator!.waveTimeoutBetweenWaves / 1000,
+    );
+    setTimeout(() => {
+      this.waveGenerator!.waveParams.isWaveInProgress = false;
+      this.waveGenerator!.isUICountdown = false;
+    }, this.waveGenerator!.waveTimeoutBetweenWaves);
+  }
+
+  public clearMemory() {
+    this.enemies = [];
+    this.projectiles = [];
+    for (let tower of this.towers!) {
+      tower.target = null;
+    }
+  }
+
+  public manageHotkeys(e: KeyboardEvent) {
+    // cancel building mode
+    if (e.key === "Escape") {
+      if (this.isCanBuild) {
+        this.isCanBuild = false;
+        this.isShowGrid = false;
+      }
+    }
+
+    if (e.key === "1") {
+      this.draftTower = null;
+      this.buildFirstTower();
+    }
+    if (e.key === "2") {
+      this.draftTower = null;
+      this.buildSecondTower();
+    }
+    if (e.key === "3") {
+      this.draftTower = null;
+      this.buildThirdTower();
+    }
+    if (e.key === "4") {
+      // release all tower target
+      for (const tower of this.towers!) {
+        tower.target = null;
+      }
+    }
+    if (e.key === "5") {
+      // remove all projectiles
+      this.projectiles = [];
+    }
+    // log mode
+    if (e.key === "0") {
+      // debug
+      console.log("this.enemies");
+      console.log(this.enemies);
+      console.log(this.enemies?.length);
+      console.log("----");
+      console.log(`this.towers`);
+      console.log(this.towers);
+      console.log(this.towers?.length);
+      console.log("----");
+      console.log(`this.projectiles`);
+      console.log(this.projectiles);
+      console.log(this.projectiles?.length);
+      console.log("---");
+      console.log(`this.lives`);
+      console.log(this.lives);
+      console.log(`this.money`);
+      console.log(this.money);
+      console.log("---");
+      console.log(`this.waveGenerator.waveParams`);
+      console.log(this.waveGenerator?.waveParams);
+      console.log("---");
```

Comment:
Maybe it will be better to use single console.log call with template string? =-)
```suggestion
      console.log(`enemies:
${this.enemies}
${this.enemies?.length}
----
towers:
${this.towers}
${this.towers?.length}
----
projectiles:
${this.projectiles}
${this.projectiles?.length}
---
lives:
${this.lives}
money:
${this.money}
---
waveGenerator.waveParams:
${this.waveGenerator?.waveParams}
---`)
```

Reply:
Skipped, debug mode by "0" keyboard key should be disabled in production